client/update.go

```diff
@@ -221,11 +217,13 @@ func (c *Channel) updateGeneric(
 	return c.enableNotifyUpdate(ctx)
 }
 
-func (c *Channel) discardUpdate(ctx context.Context) (err error) {
-	if derr := c.machine.DiscardUpdate(ctx); derr != nil {
-		// discarding update should never fail
-		err = errors.WithMessagef(derr,
-			"progressing update failed: %v, then discarding update failed", err)
+func (c *Channel) handleUpdateError(ctx context.Context, updateErr error) (err error) {
+	if err != nil {
+		if derr := c.machine.DiscardUpdate(ctx); derr != nil {
+			// discarding update should never fail
+			err = errors.WithMessagef(derr,
+				"progressing update failed: %v, then discarding update failed", updateErr)
```

Comment:
```suggestion
                       if derr == nil {
                       return err
                       }
			return errors.WithMessagef(updateErr, "progressing update failed, then discarding update failed", derr)
```

In it's current form, when `derr` is `nil`, `errors.WithMessagef` will also return a nil. So `updateErr` will be missed even if it is not nil.

Separated into two returns because, when `derr` is nil, the phrase "then discarding update failed" will be confusing.

Reply:
@matthiasgeihs

Reply:
derr cannot be nil there because `; derr != nil {`

Reply:
Sorry, i missed that part. Then, it can be as it was before.

However, I think you may have to do `errors.Messagef(updateErr..` instead of `errors.Messagef(derr..`  Because if only the message of `updateErr` is included, I think it cannot be retrieved later using `errors.As`.

Reply:
OK. Changing the ordering seemed unclean. Instead I do according to your first suggestion, don't change the error, just log `derr`. Please have a look.